Link unshortener fails to shorten several URLs

<!-- Before filing an issue, please search the existing issues (both open and closed) to see if your report might be duplicate -->
<!-- Please don't file an issue when you have a support question, send support questions to the mailinglist or ask them on Gitter (http://gitter.im/loklak/loklak) -->

- Issue type: Bug report<!-- delete the one that does not apply -->

### Short description
<!-- Explain in a few sentences what the issue/request is -->
Loklak fails to unshorten many URLs, most of them are of type `https://fb.me/*`.
<!--
If this is a bug report, use the following part of the template and delete the part at the bottom
-->
### Environment
<!-- Tell us about the environment -->
 - Operating system: CentOS 7
 - Software version: Even with `loklak:development`.
 - Software source: Github Repository<!-- e.g. Github Repository, Downloaded Jar, installed via package managers -->

### Steps to reproduce
<!-- Tell us step-by-step how the issue can be triggered. Attach any screenshots if available -->
1. Run loklak
```bash
$ ./bin/start.sh
```
2. Look at the log
```bash
$ tail -f data/loklak.log | less
```
3. You can see that unshorten fails for many URLs.

### Expected behaviour
The links should be shortened.

### Actual behaviour
Shortening links fail.

From what I can see, the unshortening fails because of two major reasong - 

### 1. Redirects with non 301 status code
Most of the fail logs are coming because of [this](https://github.com/loklak/loklak_server/commit/9b4a7585bc893b064aaf4ac79119ca1c842fd269) -
```java
            if (httpResponse.getStatusLine().getStatusCode() == 301) {
                for (Header header: httpResponse.getAllHeaders()) {
                    if (header.getName().toLowerCase().equals("location")) return header.getValue();
                }
                throw new IOException("redirect for  " + urlstring+ ": no location attribute found");
            } else {
                throw new IOException("no redirect for  " + urlstring+ " fail: " + httpResponse.getStatusLine().getStatusCode() + ": " + httpResponse.getStatusLine().getReasonPhrase());
            }
```
 It throws an exception when status code is not 301.

Though the implementation is correct, it gives a false alarm for the logger. It can simply return the original URL in the case and it will do fine.

The URLs which do not give 301 but need redirection lead to webpages which set `http-equiv="refresh"` meta tag to redirect. e.g. -
```html
$ curl http://okt.to/CV08b4
<!DOCTYPE HTML>
<html>
<head>
...
	<meta http-equiv="refresh" content="2;url=https://www.marketingweek.com/2017/04/19/does-digital-culture-exist/">
...
</head>
<body>
...
</body>
</html>
```
In such cases, we can get the unshortened URL from the HTML content of the response.

@Orbiter: Shall I try to fetch the URL from HTML in case of 2XX status code?

### 2. `SSLHandshakeException`
I noticed this type of stack trace while I was connected to my college's network - 
```java
javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1002) ~[?:1.8.0_131]
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1385) ~[?:1.8.0_131]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1413) ~[?:1.8.0_131]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1397) ~[?:1.8.0_131]
        at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:394) ~[loklak_server-all.jar:?]
        at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82) ~[loklak_server-all.jar:?]
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107) ~[loklak_server-all.jar:?]
        at org.loklak.http.ClientConnection.getRedirect(ClientConnection.java:261) ~[loklak_server-all.jar:?]
        at org.loklak.http.ClientConnection.getRedirect(ClientConnection.java:291) ~[loklak_server-all.jar:?]
        at org.loklak.harvester.RedirectUnshortener.unShorten(RedirectUnshortener.java:81) [loklak_server-all.jar:?]
        at org.loklak.harvester.TwitterScraper.unshorten(TwitterScraper.java:559) [loklak_server-all.jar:?]
        at org.loklak.harvester.TwitterScraper$TwitterTweet.run(TwitterScraper.java:489) [loklak_server-all.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
Caused by: java.io.EOFException: SSL peer shut down incorrectly
        at sun.security.ssl.InputRecord.read(InputRecord.java:505) ~[?:1.8.0_131]
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:983) ~[?:1.8.0_131]
        ... 23 more
```
I think that these are due to Cyberoam and firewall things. These may be resolved by using proper proxy settings.
```java
System.setProperty("http.proxyHost", "proxy.com");
System.setProperty("http.proxyPort", "911");
```

--- a/src/org/loklak/harvester/RedirectUnshortener.java
+++ b/src/org/loklak/harvester/RedirectUnshortener.java
@@ -75,17 +75,22 @@ public class RedirectUnshortener {
     
     public static String unShorten(String urlstring) {
         //long start = System.currentTimeMillis();
+        if (!isApplicable(urlstring)) {
+            return urlstring;
+        }
         try {
             int termination = 10; // loop for recursively shortened urls
-            while (isApplicable(urlstring) && termination-- > 0) {
+            while (termination-- > 0) {
                 String unshortened = ClientConnection.getRedirect(urlstring);
-                if (unshortened.equals(urlstring)) return urlstring;
+                if (unshortened.equals(urlstring)) {
+                    return urlstring;
+                }
                 urlstring = unshortened; // recursive apply unshortener because some unshortener are applied several times
             }
             //DAO.log("UNSHORTENED in " + (System.currentTimeMillis() - start) + " milliseconds: " + urlstring);
             return urlstring;
         } catch (IOException e) {
-            DAO.log("UNSHORTEN failed for " + urlstring);
+            DAO.log("UNSHORTEN failed for " + urlstring + " : " + e);
             return urlstring;
         }
     }

--- a/src/org/loklak/http/ClientConnection.java
+++ b/src/org/loklak/http/ClientConnection.java
@@ -6,12 +6,12 @@
  *  modify it under the terms of the GNU Lesser General Public
  *  License as published by the Free Software Foundation; either
  *  version 2.1 of the License, or (at your option) any later version.
- *  
+ *
  *  This library is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  *  Lesser General Public License for more details.
- *  
+ *
  *  You should have received a copy of the GNU Lesser General Public License
  *  along with this program in the file lgpl21.txt
  *  If not, see <http://www.gnu.org/licenses/>.
@@ -21,9 +21,11 @@ package org.loklak.http;
 
 import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
+import java.io.BufferedReader;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.FileOutputStream;
+import java.io.InputStreamReader;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.net.SocketTimeoutException;
@@ -64,20 +66,20 @@ import org.apache.http.impl.client.HttpClients;
 import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
 import org.apache.http.ssl.SSLContextBuilder;
 import org.apache.http.util.EntityUtils;
-import org.eclipse.jetty.util.log.Log;
 import org.loklak.data.DAO;
 
 /**
  * Helper class to provide BufferedReader Objects for get and post connections
  */
 public class ClientConnection {
-    
+
     public static String USER_AGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.98 Safari/537.36";
-    
+
     public  static final String CHARSET = "UTF-8";
     private static final byte LF = 10;
     private static final byte CR = 13;
     public static final byte[] CRLF = {CR, LF};
+    private static final boolean debugLog = true;
 
     public static PoolingHttpClientConnectionManager cm;
     private static RequestConfig defaultRequestConfig = RequestConfig.custom()
@@ -87,14 +89,14 @@ public class ClientConnection {
             .setContentCompressionEnabled(true)
             .setCookieSpec(CookieSpecs.IGNORE_COOKIES)
             .build();
-    
+
     private int status;
     public BufferedInputStream inputStream;
     private Map<String, List<String>> header;
     private CloseableHttpClient httpClient;
     private HttpRequestBase request;
     private HttpResponse httpResponse;
-    
+
     private static class TrustAllHostNameVerifier implements HostnameVerifier {
 		public boolean verify(String hostname, SSLSession session) {
 			return true;
@@ -117,7 +119,7 @@ public class ClientConnection {
         this.request.setHeader("User-Agent", USER_AGENT);
         this.init();
     }
-    
+
     /**
      * GET request
      * @param urlstring
@@ -126,13 +128,13 @@ public class ClientConnection {
     public ClientConnection(String urlstring) throws IOException {
     	this(urlstring, true);
     }
-    
+
     /**
      * POST request
      * @param urlstring
      * @param map
      * @param useAuthentication
-     * @throws ClientProtocolException 
+     * @throws ClientProtocolException
      * @throws IOException
      */
     public ClientConnection(String urlstring, Map<String, byte[]> map, boolean useAuthentication) throws ClientProtocolException, IOException {
@@ -141,7 +143,7 @@ public class ClientConnection {
 			.setConnectionManager(getConnctionManager(useAuthentication))
 			.setDefaultRequestConfig(defaultRequestConfig)
 			.build();
-        this.request = new HttpPost(urlstring);        
+        this.request = new HttpPost(urlstring);
         MultipartEntityBuilder entityBuilder = MultipartEntityBuilder.create();
         entityBuilder.setMode(HttpMultipartMode.BROWSER_COMPATIBLE);
         for (Map.Entry<String, byte[]> entry: map.entrySet()) {
@@ -151,7 +153,7 @@ public class ClientConnection {
         this.request.setHeader("User-Agent", USER_AGENT);
         this.init();
     }
-    
+
     /**
      * POST request
      * @param urlstring
@@ -162,14 +164,14 @@ public class ClientConnection {
     public ClientConnection(String urlstring, Map<String, byte[]> map) throws ClientProtocolException, IOException {
     	this(urlstring, map, true);
     }
-    
+
     private static PoolingHttpClientConnectionManager getConnctionManager(boolean useAuthentication){
-        
+
     	// allow opportunistic encryption if needed
-    	
+
     	boolean trustAllCerts = !"none".equals(DAO.getConfig("httpsclient.trustselfsignedcerts", "peers"))
     			&& (!useAuthentication || "all".equals(DAO.getConfig("httpsclient.trustselfsignedcerts", "peers")));
-    	
+
     	Registry<ConnectionSocketFactory> socketFactoryRegistry = null;
     	if(trustAllCerts){
 	    	try {
@@ -182,25 +184,25 @@ public class ClientConnection {
 		                .register("https", trustSelfSignedSocketFactory)
 		                .build();
 			} catch (KeyManagementException | NoSuchAlgorithmException | KeyStoreException e) {
-				Log.getLog().warn(e);
+				DAO.severe(e);
 			}
     	}
-        
-    	PoolingHttpClientConnectionManager cm = (trustAllCerts && socketFactoryRegistry != null) ? 
+
+    	PoolingHttpClientConnectionManager cm = (trustAllCerts && socketFactoryRegistry != null) ?
         		new PoolingHttpClientConnectionManager(socketFactoryRegistry):
         		new PoolingHttpClientConnectionManager();
-    	
+
         // twitter specific options
         cm.setMaxTotal(200);
         cm.setDefaultMaxPerRoute(20);
         HttpHost twitter = new HttpHost("twitter.com", 443);
         cm.setMaxPerRoute(new HttpRoute(twitter), 50);
-        
+
         return cm;
     }
 
     private void init() throws IOException {
-    	
+
         this.httpResponse = null;
         try {
             this.httpResponse = httpClient.execute(this.request);
@@ -241,44 +243,138 @@ public class ClientConnection {
             throw new IOException("client connection to " + this.request.getURI() + " fail: no connection");
         }
     }
-    
+
+    public static CloseableHttpClient getCustomClosableHttpClient(boolean useAuthentication) {
+        return HttpClients.custom()
+                .setConnectionManager(getConnctionManager(useAuthentication))
+                .setDefaultRequestConfig(defaultRequestConfig)
+                .build();
+    }
+
+    /**
+     * Get body of HttpEntity as String
+     * @param httpREntity
+     * @return response body
+     * @throws IOException when readLine fails
+     */
+    public static String getHTML(HttpEntity httpEntity) throws IOException {
+        StringBuilder sb = new StringBuilder();
+        BufferedReader reader = new BufferedReader(new InputStreamReader(httpEntity.getContent()));
+        String content = null;
+        while ((content = reader.readLine()) != null)
+            sb.append(content);
+        return sb.toString();
+    }
+
+    /**
+     * Get redirect URL from Meta tag in response body
+     * @param httpEntity
+     * @return redirect URL if any else null
+     * @throws IOException when getHTML() fails
+     */
+    public static String getMetaRedirectURL(HttpEntity httpEntity) throws IOException {
+        String html = getHTML(httpEntity);
+        html = html.replace("\n", "");  // May have a line break between tags
+        if (html.length() == 0)
+            return null;
+        int indexHttpEquiv = html.toLowerCase().indexOf("http-equiv=\"refresh\"");  // Converting to lowercase because of case insensitivity
+        if (indexHttpEquiv < 0) {
+            return null;
+        }
+        html = html.substring(indexHttpEquiv);
+        int indexContent = html.toLowerCase().indexOf("content=");
+        if (indexContent < 0) {
+            return null;
+        }
+        html = html.substring(indexContent);
+        int indexURLStart = html.toLowerCase().indexOf(";url=");
+        if (indexURLStart < 0) {
+            return null;
+        }
+        html = html.substring(indexURLStart + 5);
+        int indexURLEnd = html.toLowerCase().indexOf("\"");
+        if (indexURLEnd < 0) {
+            return null;
+        }
+        return html.substring(0, indexURLEnd);
+    }
+
+    /**
+     * Find value of location header from the given HttpResponse
+     * @param httpResponse
+     * @return Value of location field if exists, null otherwise
+     */
+    public static String getLocationHeader(HttpResponse httpResponse) {
+        for (Header header: httpResponse.getAllHeaders()) {
+            if (header.getName().equalsIgnoreCase("location")) {
+                return header.getValue();
+            }
+        }
+        return null;
+    }
+
     /**
      * get a redirect for an url: this method shall be called if it is expected that a url
      * is redirected to another url. This method then discovers the redirect.
      * @param urlstring
      * @param useAuthentication
+     * @param isGet
      * @return the redirect url for the given urlstring
-     * @throws IOException if the url is not redirected
+     * @throws IOException if there is some error with the link
      */
-    public static String getRedirect(String urlstring, boolean useAuthentication) throws IOException {
-        HttpGet get = new HttpGet(urlstring);
-        get.setConfig(RequestConfig.custom().setRedirectsEnabled(false).build());
-        get.setHeader("User-Agent", USER_AGENT);
-        CloseableHttpClient httpClient = HttpClients.custom()
-        		.setConnectionManager(getConnctionManager(useAuthentication))
-        		.setDefaultRequestConfig(defaultRequestConfig)
-        		.build();
-        HttpResponse httpResponse = httpClient.execute(get);
+    public static String getRedirect(String urlstring, boolean useAuthentication, boolean isGet) throws IOException {
+        HttpRequestBase req;
+        if (isGet)
+            req = new HttpGet(urlstring);
+        else
+            req = new HttpPost(urlstring);
+        req.setConfig(RequestConfig.custom().setRedirectsEnabled(false).build());
+        req.setHeader("User-Agent", USER_AGENT);
+        CloseableHttpClient httpClient = getCustomClosableHttpClient(useAuthentication);
+        HttpResponse httpResponse = httpClient.execute(req);
         HttpEntity httpEntity = httpResponse.getEntity();
+        if (debugLog) {
+            DAO.log("[" + (isGet ? "GET": "POST") + "] Status for " + urlstring + ": " + httpResponse.getStatusLine().toString());
+        }
         if (httpEntity != null) {
-            if (httpResponse.getStatusLine().getStatusCode() == 301) {
-                for (Header header: httpResponse.getAllHeaders()) {
-                    if (header.getName().equalsIgnoreCase("location")) {
-                        EntityUtils.consumeQuietly(httpEntity);
-                        return header.getValue();
-                    }
+            int httpStatusCode = httpResponse.getStatusLine().getStatusCode();
+            if (300 <= httpStatusCode && httpStatusCode <= 308) {
+                String location = getLocationHeader(httpResponse);
+                EntityUtils.consumeQuietly(httpEntity);
+                if (location != null) {  // A redirect was found
+                    return location;
                 }
-                EntityUtils.consumeQuietly(httpEntity);
                 throw new IOException("redirect for  " + urlstring+ ": no location attribute found");
+            } else if (isGet && httpStatusCode != 200) {
+                /*
+                    In most of the cases when requesting to http://fb.me/*, a GET request results in 400 Bad Request
+                    But making a POST request does the job.
+                */
+                if (debugLog) {
+                    DAO.log("GET method failed for " + urlstring + " trying POST");
+                }
+                return getRedirect(urlstring, useAuthentication, false);
             } else {
+                if (debugLog)
+                    DAO.log("Trying to fetch Meta redirect URL for " + urlstring);
+                String metaURL = getMetaRedirectURL(httpEntity);
                 EntityUtils.consumeQuietly(httpEntity);
-                throw new IOException("no redirect for  " + urlstring+ " fail: " + httpResponse.getStatusLine().getStatusCode() + ": " + httpResponse.getStatusLine().getReasonPhrase());
+                if (metaURL != null) {  // A URL was found in meta tag
+                    if (debugLog) {
+                        DAO.log("Fetched Meta redirect URL for " + urlstring + " : " + metaURL);
+                    }
+                    if (!metaURL.startsWith("http") && !urlstring.endsWith("/")) {
+                        return urlstring + "/" + metaURL;
+                    }
+                    return metaURL;
+                }
+                return urlstring;
             }
         } else {
             throw new IOException("client connection to " + urlstring + " fail: no connection");
         }
     }
-    
+
     /**
      * get a redirect for an url: this method shall be called if it is expected that a url
      * is redirected to another url. This method then discovers the redirect.
@@ -287,9 +383,13 @@ public class ClientConnection {
      * @throws IOException
      */
     public static String getRedirect(String urlstring) throws IOException {
-    	return getRedirect(urlstring, true);
+        return getRedirect(urlstring, true, true);
     }
-    
+
+    public static String getRedirect(String urlstring, boolean useAuthentication) throws IOException {
+        return getRedirect(urlstring, useAuthentication, true);  // Try GET request first
+    }
+
     public void close() {
         HttpEntity httpEntity = this.httpResponse.getEntity();
         if (httpEntity != null) EntityUtils.consumeQuietly(httpEntity);
@@ -299,7 +399,7 @@ public class ClientConnection {
             this.request.releaseConnection();
         }
     }
-    
+
     public static void download(String source_url, File target_file, boolean useAuthentication) {
         try {
             ClientConnection connection = new ClientConnection(source_url, useAuthentication);
@@ -310,28 +410,28 @@ public class ClientConnection {
                 try {
                     while ((count = connection.inputStream.read(buffer)) > 0) os.write(buffer, 0, count);
                 } catch (IOException e) {
-                	Log.getLog().warn(e.getMessage());
+                	DAO.severe(e);
                 } finally {
                     os.close();
                 }
             } catch (IOException e) {
-            	Log.getLog().warn(e.getMessage());
+            	DAO.severe(e);
             } finally {
                 connection.close();
             }
         } catch (IOException e) {
-        	Log.getLog().warn(e.getMessage());
+        	DAO.severe(e);
         }
     }
-    
+
     public static void download(String source_url, File target_file) {
     	download(source_url, target_file, true);
     }
-    
+
     public static void downloadPeer(String source_url, File target_file) {
     	download(source_url, target_file, !"peers".equals(DAO.getConfig("httpsclient.trustselfsignedcerts", "peers")));
     }
-    
+
     public static byte[] download(String source_url, boolean useAuthentication) throws IOException {
         try {
             ClientConnection connection = new ClientConnection(source_url);
@@ -342,21 +442,21 @@ public class ClientConnection {
             try {
                 while ((count = connection.inputStream.read(buffer)) > 0) baos.write(buffer, 0, count);
             } catch (IOException e) {
-            	Log.getLog().warn(e.getMessage());
+            	DAO.severe(e);
             } finally {
                 connection.close();
             }
             return baos.toByteArray();
         } catch (IOException e) {
-        	Log.getLog().warn(e.getMessage());
+        	DAO.severe(e);
             return null;
         }
     }
-    
+
     public static byte[] download(String source_url) throws IOException {
     	return download(source_url, true);
     }
-    
+
     public static byte[] downloadPeer(String source_url) throws IOException {
     	return download(source_url, !"peers".equals(DAO.getConfig("httpsclient.trustselfsignedcerts", "peers")));
     }

